training/worker.py

```diff
@@ -66,8 +71,10 @@ def ForwardMode(self, request, context):
         )
 
         start = time.time()
-        ctxt.forward_mode()
 
+        # ensure only one GPU can be running at given time.
+        self.mutex.acquire()
```

Comment:
Prob doesn’t matter here but

try:
  mutex.acquire()
finally:
  mutex.release()

Will prevent errors holding locks

Reply:
I think i'd want the error to be raised and propagated back to the host

Reply:
It will be, The finally block just ensures the mutex is released

Whatever is in “finally” is always run whether an exception is thrown or not.

You could also use a “with mutex” lock idiom as well.

Reply:
Ah I see what you mean: you're concerned about code in between the acquire() and the release() throwing an exception deadlocking the code. Yeah it's a definite possibility. I'll change it to a context manager.